### PR TITLE
Implement embed directly in Outcome

### LINKED ISSUE #1232: Copy embed machinery to Outcome
`Concurrent#background` is a little annoying right now because it returns `F[Outcome[F, E, A]]` in its body. This is consistent with `join`, so that makes sense, but `join` also has `joinAndEmbed` and related, and there's no analogue of this for `Outcome`. It might make sense to add `embed` as a method on `Outcome` to help ease this a bit. The `joinAndEmbed` machinery could then delegate to this method + `flatten`.

Open for discussion. :-) Could also have a different variant of `background`.

### PR DESCRIPTION
**Pull request by vasilmkd**
Resolves #1232.

### CODE REVIEW

File: kernel/shared/src/main/scala/cats/effect/kernel/Fiber.scala [hunk 1]
```diff
@@ -22,8 +22,8 @@ trait Fiber[F[_], E, A] {
   def cancel: F[Unit]
   def join: F[Outcome[F, E, A]]
 
-  def joinAndEmbed(onCancel: F[A])(implicit F: MonadCancel[F, E]): F[A] =
-    join.flatMap(_.fold(onCancel, F.raiseError(_), fa => fa))
+  def joinAndEmbed(onCancel: => F[A])(implicit F: MonadCancel[F, E]): F[A] =
```
Comment (djspiewak) [inline kernel/shared/src/main/scala/cats/effect/kernel/Fiber.scala]: Why by-name?

File: kernel/shared/src/main/scala/cats/effect/kernel/GenSpawn.scala [hunk 2]
```diff
@@ -108,6 +108,12 @@ trait GenSpawn[F[_], E] extends MonadCancel[F, E] {
 
   def background[A](fa: F[A]): Resource[F, F[Outcome[F, E, A]]] =
     Resource.make(start(fa))(_.cancel)(this).map(_.join)(this)
+
+  def backgroundAndEmbed[A](fa: F[A])(onCancel: => F[A]): Resource[F, F[A]] =
+    background(fa).map(flatMap(_)(_.embed(onCancel)(this)))(this)
+
+  def backgroundAndEmbedNever[A](fa: F[A]): Resource[F, F[A]] =
+    backgroundAndEmbed(fa)(never)
```
Comment (djspiewak) [inline kernel/shared/src/main/scala/cats/effect/kernel/GenSpawn.scala]: Discussion: is it worth having this, or should we just rely on people mapping over the contents? Like, `background(fa).evalMap(_.embedNever)` doesn't seem too onerous.

File: kernel/shared/src/main/scala/cats/effect/kernel/Outcome.scala [hunk 3]
```diff
@@ -26,6 +26,12 @@ import scala.util.{Either, Left, Right}
 sealed trait Outcome[F[_], E, A] extends Product with Serializable {
   import Outcome._
 
+  def embed(onCancel: => F[A])(implicit F: MonadCancel[F, E]): F[A] =
```
Comment (djspiewak) [inline kernel/shared/src/main/scala/cats/effect/kernel/Outcome.scala]: Why by-name?

**Review by djspiewak — CHANGES_REQUESTED** [review]

File: kernel/shared/src/main/scala/cats/effect/kernel/Fiber.scala (hunk 1 above)
Comment (vasilmkd) [reply kernel/shared/src/main/scala/cats/effect/kernel/Fiber.scala]: `Outcome#fold` takes a by-name parameter. I would think the `onCancel` does not need to be fully built if that path is not taken? Is that a wrong assumption?

Comment (djspiewak) [reply kernel/shared/src/main/scala/cats/effect/kernel/Fiber.scala]: That's definitely a correct assumption. I think my thought on the `fold` signature is the fact that it's more general, and critically, people might recurse through it (think: just like `whenM` and `>>`). I can't really see people making `joinAndEmbed` the branch point in a conditional recursion, but maybe I'm wrong on that one.

Comment (vasilmkd) [reply kernel/shared/src/main/scala/cats/effect/kernel/Fiber.scala]: I'll revert to strict.

File: kernel/shared/src/main/scala/cats/effect/kernel/GenSpawn.scala (hunk 2 above)
Comment (vasilmkd) [reply kernel/shared/src/main/scala/cats/effect/kernel/GenSpawn.scala]: I'm not a fan of the names, but I do like the simmetry with `join`.

File: kernel/shared/src/main/scala/cats/effect/kernel/Outcome.scala (hunk 3 above)
Comment (vasilmkd) [reply kernel/shared/src/main/scala/cats/effect/kernel/Outcome.scala]: Reverted.

File: kernel/shared/src/main/scala/cats/effect/kernel/GenSpawn.scala (hunk 2 above)
Comment (vasilmkd) [reply kernel/shared/src/main/scala/cats/effect/kernel/GenSpawn.scala]: I removed these methods.

File: kernel/shared/src/main/scala/cats/effect/kernel/Fiber.scala (hunk 1 above)
Comment (vasilmkd) [reply kernel/shared/src/main/scala/cats/effect/kernel/Fiber.scala]: Reverted.

**Comment by djspiewak** [comment]
🎉

**Pull request merged**